src/lib.rs

```diff
@@ -2361,29 +2361,42 @@ impl Build {
                     let mut parts = target.split('-');
                     if let Some(arch) = parts.next() {
                         let arch = &arch[5..];
+
+                        // Assume that "rv{arch}" is a valid RISC-V ISA string.
+                        // The compiler would error out otherwise, and we fix
+                        // that later.
+                        cmd.args.push(format!("-march=rv{arch}").into());
+
+                        // Detect single-letter extensions from `arch`, assuming
+                        // no version numbers and canonical order
+                        let single_letter = arch
+                            .split(['_', 'z', 's'])
+                            .next()
+                            // The arch string starts with 32 or 64
+                            .expect("arch string cannot be empty");
```

Comment:
I'm bit confused on this, I checked the riscv targets rust supported and doesn't find any z, s or underscore in after the prefix "riscv" .

What is this trying to detect?
Is there any new target which contains z, s or underscore?

Reply:
i don't know how else to address the fact that isa string parsing is "fragile" https://github.com/rust-lang/cc-rs/pull/796#discussion_r1136891753

in case in the future we add extensions like `zsomething` the arch string could look like `riscv64imaczsomething` or `riscv64imac_zsomething` and we don't want to match on the `g` of `zsomething`

i don't think there's any easy way to say how the rust project will decide to name future riscv architectures, but if you're okay with this being "fragile" i can remove this part

Reply:
just to chim in here, to give an idea about how for example the riscv-esp-elf-gcc sees target definitions, it looks something like the following:

```
*multilib_defaults:
march=rv32imafdc_zicsr_zifencei mabi=ilp32

*multilib_extra:


*multilib_matches:
march=rv32i_zicsr_zifencei march=rv32i_zicsr_zifencei;march=rv32iac_zicsr_zifencei march=rv32iac_zicsr_zifencei;march=rv32ic_zicsr_zifencei march=rv32ic_zicsr_zifencei;march=rv32im_zicsr_zifencei march=rv32im_zicsr_zifencei;march=rv32imc_zicsr_zifencei march=rv32imc_zicsr_zifencei;march=rv32imac_zicsr_zifencei march=rv32imac_zicsr_zifencei;march=rv32imafc_zicsr_zifencei march=rv32imafc_zicsr_zifencei;march=rv32imafdc_zicsr_zifencei march=rv32imafdc_zicsr_zifencei;march=rv32imafdc_zicsr_zifencei_xesppie march=rv32imafdc_zicsr_zifencei_xesppie;mabi=ilp32 mabi=ilp32;mabi=ilp32f mabi=ilp32f;fno-rtti fno-rtti;
```

this above is part of the output of `./riscv32-esp-elf-gcc -dumpspecs` and presumably it looks simialr on other variants.

Reply:
Oh so it's non-standard rust target used in riscv-esp-elf-gcc?

If it is non-standard then it would indeed makes more sense to pass them as environment variables.

Reply:
I am not sure what you mean by non-standard rust target, but [our target descriptions](https://github.com/SergioGasquez/rust/blob/b092798128a9c957952037638ea6e834be405e13/src/doc/rustc/src/platform-support/esp-idf.md#requirements) are similar on how the other riscv targets on the rust site are being expressed. 

But as dramforever mentions downstream gcc & clang is more [complicated](https://github.com/SergioGasquez/rust/blob/b092798128a9c957952037638ea6e834be405e13/src/doc/rustc/src/platform-support/esp-idf.md#requirements) matter. Mapping to thouse means that there are possible different ISA versions that can be associated with. Rust target names in its current form lag that expressiveness if i see it correctly.

Reply:
> I am not sure what you mean by non-standard rust target, but our target descriptions are similar on how the other riscv targets on the rust site are being expressed.

I consider any target not in rustup or maintainer by the rust officially to be a non-standard one.

I heard of cargo plugins for esp and I think it sets the cflags and etc to work?

Reply:
I guess it might make sense to summarize the situation:

- Rust targets lists the instruction extensions after `riscv{32,64}`. `f` means hardware f32, `d` means hardware f64, and `g` is abbreviation of `imafd`
- *Currently* there are only single letters there so we can detect it based on just looking for `g`/`d` or `f`
- Previously a maintainer complained that it was fragile... but with no real way to predict future target names IMO it will never be not fragile
- There's no way to get the target ABI used by rustc here, IIUC. If there was this whole detection would be obsolete for everyone.
- The de facto standard for `-mabi` is to use the "maximum" supported hard float ABI

Reply:
> There's no way to get the target ABI used by rustc here, IIUC. If there was this whole detection would be obsolete for everyone.

We do have a crate to extract some riscv target information from nightly rust.

I wonder if we could get this from it as well?

https://github.com/rust-lang/cc-rs/blob/main/dev-tools/gen-target-info/src/main.rs